src/_common/reaction/reaction-count.ts

```diff
@@ -4,18 +4,122 @@ import { getCurrentServerTime } from '../../utils/server-time';
 
-export abstract class ReactionableModel extends Model {
-	declare reaction_counts: ReactionCount[];
+export interface EmojiDelta {
+	emoji_id: number;
+	emoji_short_name: string;
+	emoji_prefix: string;
+	emoji_img_url: string;
 
-	get typename__(): string {
-		throw new Error('ReactionableModel must implement typename__');
+	// user ids which reacted / unreacted to this emoji
+	delta_inc: number[];
+	delta_dec: number[];
+}
+
+export interface ReactionableModel {
+	id: number;
+	reaction_counts: ReactionCount[];
+
+	get resourceName(): string;
+}
+
+interface EmojiIdAndCount {
+	emoji_id: number;
+	count: number;
+}
+const ReactionQueue = new WeakMap<ReactionableModel, EmojiIdAndCount[]>();
+
+function applyReaction(
+	model: ReactionableModel,
+	existingReaction: ReactionCount | undefined,
+	countMod: number,
+	emojiId: number,
+	emojiImgUrl: string,
+	emojiPrefix: string,
+	emojiShortName: string,
+	didReact: boolean
+) {
+	if (existingReaction) {
+		if (existingReaction.count + countMod <= 0) {
+			arrayRemove(model.reaction_counts, i => i.id === emojiId);
+		} else {
+			existingReaction.count = existingReaction.count + countMod;
+			existingReaction.did_react = didReact;
+		}
+	} else if (countMod > 0) {
+		model.reaction_counts.push(
+			new ReactionCount({
+				id: emojiId,
+				img_url: emojiImgUrl,
+				prefix: emojiPrefix,
+				short_name: emojiShortName,
+				count: countMod,
+				did_react: didReact,
+			})
+		);
 	}
 }
 
+export async function updateReactionCountForAnEmoji(
+	model: ReactionableModel,
+	emojiId: number,
+	emojiShortName: string,
+	emojiPrefix: string,
+	emojiImgUrl: string,
+	deltaInc: number[],
+	deltaDec: number[],
+	current_user_id: number
+) {
+	let countMod = deltaInc.length - deltaDec.length;
+
+	const existingReaction = model.reaction_counts.find(i => i.id === emojiId);
+	const user_reacting_count = deltaInc.filter(id => id === current_user_id).length;
+	const user_unreacting_count = deltaDec.filter(id => id === current_user_id).length;
+
+	let didReact = user_reacting_count > 0;
+	let updateDidReact = false;
+
+	if (existingReaction) {
+		// emoji.did_react will need to be updated if user_reacting_count and user_unreacting_count are not equal
+		if (user_reacting_count != user_unreacting_count) {
+			didReact = !existingReaction.did_react;
+			updateDidReact = true;
+		} else {
+			didReact = existingReaction.did_react;
+		}
+	}
+
+	// we need to cancel out the self reactions (applied earlier) from the payload we receive
+	const modelSelfReactionQueue = ReactionQueue.get(model) ?? [];
+	if (modelSelfReactionQueue.length) {
+		const emojiSelfReactionQueue = modelSelfReactionQueue.find(i => i.emoji_id === emojiId);
+		if (emojiSelfReactionQueue) {
+			countMod = countMod - emojiSelfReactionQueue.count;
+			arrayRemove(modelSelfReactionQueue, i => i.emoji_id === emojiId);
+			ReactionQueue.set(model, modelSelfReactionQueue);
+			updateDidReact = false;
```

Comment:
is a flag used to indicate that  state has been change and requires an ui update. 
it's being set to  here because having the pending reactions here shows that this client has already made the  highlight update and we don't have to update the UI for the state change.

I've renamed this var to `updateDidReactIndicator`  and added comment explaining this.